### PR TITLE
Only save StudentImage if the face is inside the frame [4-8h] #148

### PR DESCRIPTION
**Pull request by sladomic**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -91,6 +91,6 @@ dependencies {
     compile 'pl.droidsonroids.gif:android-gif-drawable:1.2.0'
     compile 'com.skyfishjy.ripplebackground:library:1.0.1'
 
-    compile 'ch.zhaw:facerecognitionlibrary:1.2.9.3'
+    compile 'ch.zhaw:facerecognitionlibrary:1.3.0'
```
Comment (jo-elimu) [inline app/build.gradle:94]: Is this the version with the custom classifier or the original one?

File: app/src/main/assets/AnimalOverlaysConfig.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+	"animalOverlays" : {
```
Comment (jo-elimu) [inline app/src/main/assets/AnimalOverlaysConfig.json:2]: Change to `"animalOverlays" : [` (array)?

Comment (jo-elimu) [inline app/src/main/assets/AnimalOverlaysConfig.json:2]: Change to `"animalOverlays" : [`?

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java [hunk 3]
```diff
@@ -94,21 +93,32 @@ public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
...
+                    faces = MatOperation.rotateFaces(imgRgba, faces, ppF.getAngleForRecognition());
+                    face = faces[0];
+                    faceDetected = true;
+                    isFaceInsideFrame = DetectionHelper.isFaceInsideFrame(animalOverlay, imgRgba, face);
+
+                    if (isFaceInsideFrame){
+                        String svmString = getSvmString(img);
+                        String svmProbability = svm.recognizeProbability(svmString);
+                        Student student = getStudentFromProbability(svmProbability);
+                        numberOfTries++;
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:116]: Add logging of number of tries to make testing easier.

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java [hunk 4]
```diff
@@ -117,6 +127,10 @@ public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
         // Add overlay
         animalOverlayHelper.addOverlay(imgRgba);
 
+        if (faceDetected && !isFaceInsideFrame){
+            DetectionHelper.drawArrowFromFaceToFrame(animalOverlay, imgRgba, face);
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:133]: Nice solution :-)

File: app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java [hunk 5]
```diff
@@ -49,6 +55,7 @@
     private LiteracyApplication literacyApplication;
     private List<Mat> studentImages;
     private AnimalOverlayHelper animalOverlayHelper;
+    private AnimalOverlay animalOverlay;
 
     // Image collection parameters
     private static final boolean DIAGNOSE_MODE = true;
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java:61]: Should DIAGNOSE_MODE be set to "false" before every release?

File: app/src/main/java/org/literacyapp/authentication/animaloverlay/AnimalOverlayHelper.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        this.context = context;
+        imgMask = new Mat();
+        imgInvMask = new Mat();
+    }
+
+    public AnimalOverlay createOverlay() {
+        Log.i(getClass().getName(), "createOverlay");
+        List<AnimalOverlay> animalOverlays = getAnimalOverlays();
+
+        int randomNumber = (int) (Math.random() * animalOverlays.size());
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/animaloverlay/AnimalOverlayHelper.java]: Rename to `randomIndex`?

**Review by jo-elimu — APPROVED** [review]

File: app/build.gradle (hunk 1 above)
Comment (sladomic) [reply app/build.gradle:94]: custom..but it is not set at the moment..so at the moment the original one is used (once we decide which one we use, I can set it with setCascadeClassifierForFaceDetector in the AuthenticationActivity and StudentImageCollectionActivity)

File: app/src/main/assets/AnimalOverlaysConfig.json (hunk 2 above)
Comment (sladomic) [reply app/src/main/assets/AnimalOverlaysConfig.json:2]: I had it as an array but somehow the automatic mapping didn't work then. http://stackoverflow.com/questions/16377754/parse-json-file-using-gson but if you know how to do it, I will change it. I just don't know :)

Comment (sladomic) [reply app/src/main/assets/AnimalOverlaysConfig.json:2]: see above

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java (hunk 3 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:116]: done

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java (hunk 4 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:133]: doesn't look really nice but should work for a start (maybe together with audio?)

File: app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java (hunk 5 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java:61]: I think so, yes. Or maybe false by default and if we are testing, we can set it to true on our computers?

File: app/src/main/java/org/literacyapp/authentication/animaloverlay/AnimalOverlayHelper.java (hunk 6 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/animaloverlay/AnimalOverlayHelper.java]: done

File: app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java (hunk 5 above)
Comment (jo-elimu) [reply app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java:61]: Yes, that's probably better.

**Pull request merged**